File: devops-scm-provider/devops-scm-provider-git/devops-scm-provider-tgit/src/main/java/com/tencent/devops/scm/provider/git/tgit/TGitPullRequestService.java
package com.tencent.devops.scm.provider.git.tgit;

import com.tencent.devops.scm.api.PullRequestService;
import com.tencent.devops.scm.api.enums.PullRequestState;
import com.tencent.devops.scm.api.exception.NotFoundScmApiException;
import com.tencent.devops.scm.api.pojo.Change;
import com.tencent.devops.scm.api.pojo.Comment;
import com.tencent.devops.scm.api.pojo.CommentInput;
import com.tencent.devops.scm.api.pojo.Commit;
import com.tencent.devops.scm.api.pojo.ListOptions;
import com.tencent.devops.scm.api.pojo.PullRequest;
import com.tencent.devops.scm.api.pojo.PullRequestInput;
import com.tencent.devops.scm.api.pojo.PullRequestListOptions;
import com.tencent.devops.scm.api.pojo.repository.ScmProviderRepository;
import com.tencent.devops.scm.sdk.tgit.TGitApi;
import com.tencent.devops.scm.sdk.tgit.TGitApiFactory;
import com.tencent.devops.scm.sdk.tgit.enums.TGitMergeRequestState;
import com.tencent.devops.scm.sdk.tgit.pojo.TGitCommit;
import com.tencent.devops.scm.sdk.tgit.pojo.TGitMergeRequest;
import com.tencent.devops.scm.sdk.tgit.pojo.TGitMergeRequestFilter;
import com.tencent.devops.scm.sdk.tgit.pojo.TGitMergeRequestParams;
import com.tencent.devops.scm.sdk.tgit.pojo.TGitMergeRequestParams.TGitMergeRequestParamsBuilder;
import com.tencent.devops.scm.sdk.tgit.pojo.TGitNote;
import com.tencent.devops.scm.sdk.tgit.pojo.TGitProject;
import java.util.List;
import java.util.stream.Collectors;
import org.apache.commons.lang3.StringUtils;

public class TGitPullRequestService implements PullRequestService {

    private final TGitApiFactory apiFactory;

    public TGitPullRequestService(TGitApiFactory apiFactory) {
        this.apiFactory = apiFactory;
    }

    @Override
    public PullRequest find(ScmProviderRepository repository, Integer number) {
        return TGitApiTemplate.execute(repository, apiFactory, (repo, tGitApi) -> {
            TGitMergeRequest mergeRequest = tGitApi.getMergeRequestApi()
                    .getMergeRequestByIid(repo.getProjectIdOrPath(), number);
            TGitProject targetProject = tGitApi.getProjectApi().getProject(repo.getProjectIdOrPath());
            // 授权者不一定有fork仓库的权限
            TGitProject sourceProject =
                    mergeRequest.getTargetProjectId().equals(mergeRequest.getSourceProjectId()) ? targetProject : null;
            return TGitObjectConverter.convertPullRequest(mergeRequest, sourceProject, targetProject);
        });
    }

    @Override
    public List<PullRequest> list(ScmProviderRepository repository, PullRequestListOptions opts) {
        return TGitApiTemplate.execute(repository, apiFactory, (repo, tGitApi) -> {
            TGitMergeRequestFilter filter = TGitMergeRequestFilter.builder()
                    .state(convertFromState(opts.getState()))
                    .sourceBranch(opts.getSourceBranch())
                    .targetBranch(opts.getTargetBranch())
                    .page(opts.getPage())
                    .perPage(opts.getPageSize())
                    .build();
            List<TGitMergeRequest> mergeRequests = tGitApi.getMergeRequestApi()
                    .getMergeRequests(repo.getProjectIdOrPath(), filter);
            TGitProject targetProject = tGitApi.getProjectApi().getProject(repo.getProjectIdOrPath());
            return mergeRequests.stream()
                    .map(mergeRequest ->
                            TGitObjectConverter.convertPullRequest(mergeRequest, null, targetProject)
                    )
                    .collect(Collectors.toList());
        });
    }

    @Override
    public PullRequest create(ScmProviderRepository repository, PullRequestInput input) {
        if (StringUtils.isEmpty(input.getTitle())) {
            throw new IllegalArgumentException("title cannot be empty");
        }
        if (StringUtils.isEmpty(input.getSourceBranch())) {
            throw new IllegalArgumentException("source branch cannot be empty");
        }
        if (StringUtils.isEmpty(input.getTargetBranch())) {
            throw new IllegalArgumentException("target branch cannot be empty");
        }

        return TGitApiTemplate.execute(repository, apiFactory, (repo, tGitApi) -> {
            TGitMergeRequestParamsBuilder builder = TGitMergeRequestParams.builder()
                    .title(input.getTitle())
                    .description(input.getBody())
                    .sourceBranch(input.getSourceBranch())
                    .targetBranch(input.getTargetBranch());
            if (input.getTargetRepo() != null && input.getTargetRepo() instanceof Long) {
                builder.targetProjectId((Long) input.getTargetRepo());
            }
            TGitMergeRequest mergeRequest = tGitApi.getMergeRequestApi()
                    .createMergeRequest(repo.getProjectIdOrPath(), builder.build());
            TGitProject targetProject = tGitApi.getProjectApi().getProject(mergeRequest.getTargetProjectId());
            TGitProject sourceProject =
                    mergeRequest.getTargetProjectId().equals(mergeRequest.getSourceProjectId()) ? targetProject
                            : tGitApi.getProjectApi().getProject(mergeRequest.getSourceProjectId());
            ;
            return TGitObjectConverter.convertPullRequest(mergeRequest, sourceProject, targetProject);
        });
    }

    @Override
    public List<Change> listChanges(ScmProviderRepository repository, Integer number, ListOptions opts) {
        return TGitApiTemplate.execute(repository, apiFactory, (repo, tGitApi) -> {
            TGitMergeRequest mergeRequest = getMergeRequestByIid(tGitApi, repo.getProjectIdOrPath(), number);
            TGitMergeRequest mergeRequestChanges = tGitApi.getMergeRequestApi()
                    .getMergeRequestChanges(repo.getProjectIdOrPath(), mergeRequest.getId());
            return mergeRequestChanges.getFiles().stream()
                    .map(TGitObjectConverter::convertChange)
                    .collect(Collectors.toList());
        });
    }

    @Override
    public List<Commit> listCommits(ScmProviderRepository repository, Integer number, ListOptions opts) {
        return TGitApiTemplate.execute(repository, apiFactory, (repo, tGitApi) -> {
            TGitMergeRequest mergeRequest = getMergeRequestByIid(tGitApi, repo.getProjectIdOrPath(), number);
            List<TGitCommit> mergeRequestCommits = tGitApi.getMergeRequestApi()
                    .getMergeRequestCommits(repo.getProjectIdOrPath(), mergeRequest.getId(),
                            opts.getPage(), opts.getPageSize());
            return mergeRequestCommits.stream()
                    .map(TGitObjectConverter::convertCommit)
                    .collect(Collectors.toList());
        });
    }

    @Override
    public void merge(ScmProviderRepository repository, Integer number) {
        TGitApiTemplate.execute(repository, apiFactory, (repo, tGitApi) -> {
            TGitMergeRequest mergeRequest = getMergeRequestByIid(tGitApi, repo.getProjectIdOrPath(), number);
            tGitApi.getMergeRequestApi().mergeMergeRequest(repo.getProjectIdOrPath(), mergeRequest.getId());
        });
    }

    @Override
    public void close(ScmProviderRepository repository, Integer number) {
        TGitApiTemplate.execute(repository, apiFactory, (repo, tGitApi) -> {
            TGitMergeRequest mergeRequest = getMergeRequestByIid(tGitApi, repo.getProjectIdOrPath(), number);
            tGitApi.getMergeRequestApi().closeMergeRequest(repo.getProjectIdOrPath(), mergeRequest.getId());
        });
    }

    @Override
    public Comment findComment(ScmProviderRepository repository, Integer number, Long commentId) {
        return TGitApiTemplate.execute(repository, apiFactory, (repo, tGitApi) -> {
            TGitMergeRequest mergeRequest = getMergeRequestByIid(tGitApi, repo.getProjectIdOrPath(), number);
            TGitNote mergeRequestNote = tGitApi.getNotesApi()
                    .getMergeRequestNote(repo.getProjectIdOrPath(), mergeRequest.getId(), commentId);
            return TGitObjectConverter.convertComment(mergeRequestNote);
        });
    }

    @Override
    public List<Comment> listComments(ScmProviderRepository repository, Integer number, ListOptions opts) {
        return TGitApiTemplate.execute(repository, apiFactory, (repo, tGitApi) -> {
            TGitMergeRequest mergeRequest = getMergeRequestByIid(tGitApi, repo.getProjectIdOrPath(), number);
            List<TGitNote> mergeRequestNotes = tGitApi.getNotesApi()
                    .getMergeRequestNotes(repo.getProjectIdOrPath(), mergeRequest.getId(),
                            opts.getPage(), opts.getPageSize());

            return mergeRequestNotes.stream()
                    .map(TGitObjectConverter::convertComment)
                    .collect(Collectors.toList());
        });
    }

    @Override
    public Comment createComment(ScmProviderRepository repository, Integer number, CommentInput input) {
        if (StringUtils.isBlank(input.getBody())) {
            throw new IllegalArgumentException("body cannot be empty");
        }
        return TGitApiTemplate.execute(repository, apiFactory, (repo, tGitApi) -> {
            TGitMergeRequest mergeRequest = getMergeRequestByIid(tGitApi, repo.getProjectIdOrPath(), number);
            TGitNote mergeRequestNote = tGitApi.getNotesApi()
                    .createMergeRequestNote(repo.getProjectIdOrPath(), mergeRequest.getId(), input.getBody());
            return TGitObjectConverter.convertComment(mergeRequestNote);
        });
    }

    @Override
    public void deleteComment(ScmProviderRepository repository, Integer number, Long commentId) {
    }

    private TGitMergeRequestState convertFromState(PullRequestState from) {
        if (from == null) {
            return null;
        }
        TGitMergeRequestState state;
        switch (from) {
            case OPENED:
                state = TGitMergeRequestState.OPENED;
                break;
            case REOPENED:
                state = TGitMergeRequestState.REOPENED;
                break;
            case MERGED:
                state = TGitMergeRequestState.MERGED;
                break;
            case CLOSED:
                state = TGitMergeRequestState.CLOSED;
                break;
            default:
                return null;
        }
        return state;
    }

    private TGitMergeRequest getMergeRequestByIid(TGitApi tGitApi, Object projectIdOrPath, Integer number) {
        TGitMergeRequest mergeRequest = tGitApi.getMergeRequestApi().getMergeRequestByIid(projectIdOrPath, number);
        if (mergeRequest == null) {
            throw new NotFoundScmApiException(String.format("merge request %s not found", number));
        }
        return mergeRequest;
    }
}

File: devops-scm-provider/devops-scm-provider-git/devops-scm-provider-tgit/src/main/kotlin/com/tencent/devops/scm/provider/git/tgit/TGitPullRequestService.kt
package com.tencent.devops.scm.provider.git.tgit

import com.tencent.devops.scm.api.PullRequestService
import com.tencent.devops.scm.api.enums.PullRequestState
import com.tencent.devops.scm.api.exception.NotFoundScmApiException
import com.tencent.devops.scm.api.pojo.Change
import com.tencent.devops.scm.api.pojo.Comment
import com.tencent.devops.scm.api.pojo.CommentInput
import com.tencent.devops.scm.api.pojo.Commit
import com.tencent.devops.scm.api.pojo.ListOptions
import com.tencent.devops.scm.api.pojo.PullRequest
import com.tencent.devops.scm.api.pojo.PullRequestInput
import com.tencent.devops.scm.api.pojo.PullRequestListOptions
import com.tencent.devops.scm.api.pojo.repository.ScmProviderRepository
import com.tencent.devops.scm.sdk.tgit.TGitApi
import com.tencent.devops.scm.sdk.tgit.TGitApiFactory
import com.tencent.devops.scm.sdk.tgit.enums.TGitMergeRequestState
import com.tencent.devops.scm.sdk.tgit.pojo.TGitMergeRequest
import com.tencent.devops.scm.sdk.tgit.pojo.TGitMergeRequestFilter
import com.tencent.devops.scm.sdk.tgit.pojo.TGitMergeRequestParams
import org.apache.commons.lang3.StringUtils

/**
 * TGit 合并请求服务实现类
 * @property apiFactory TGit API工厂
 */
class TGitPullRequestService(
    private val apiFactory: TGitApiFactory
) : PullRequestService {

    /**
     * 获取指定合并请求
     * @param repository 代码仓库信息
     * @param number 合并请求编号
     * @return 合并请求详情
     */
    override fun find(repository: ScmProviderRepository, number: Int): PullRequest {
        return TGitApiTemplate.execute(repository, apiFactory) { repo, tGitApi ->
            val mergeRequest = tGitApi.mergeRequestApi.getMergeRequestByIid(repo.projectIdOrPath, number)
            val targetProject = tGitApi.projectApi.getProject(repo.projectIdOrPath)
            // 授权者不一定有fork仓库的权限
            val sourceProject = if (mergeRequest.targetProjectId == mergeRequest.sourceProjectId) targetProject else null
            TGitObjectConverter.convertPullRequest(mergeRequest, sourceProject, targetProject)
        }
    }

    /**
     * 获取合并请求列表
     * @param repository 代码仓库信息
     * @param opts 列表查询选项
     * @return 合并请求列表
     */
    override fun list(repository: ScmProviderRepository, opts: PullRequestListOptions): List<PullRequest> {
        return TGitApiTemplate.execute(repository, apiFactory) { repo, tGitApi ->
            val filter = TGitMergeRequestFilter.builder()
                    .state(convertFromState(opts.state))
                    .sourceBranch(opts.sourceBranch)
                    .targetBranch(opts.targetBranch)
                    .page(opts.page)
                    .perPage(opts.pageSize)
                    .build()
            val mergeRequests = tGitApi.mergeRequestApi.getMergeRequests(repo.projectIdOrPath, filter)
            val targetProject = tGitApi.projectApi.getProject(repo.projectIdOrPath)
            mergeRequests.map { mergeRequest ->
                TGitObjectConverter.convertPullRequest(mergeRequest, null, targetProject)
            }
        }
    }

    /**
     * 创建合并请求
     * @param repository 代码仓库信息
     * @param input 合并请求输入参数
     * @return 创建的合并请求详情
     * @throws IllegalArgumentException 当标题、源分支或目标分支为空时抛出
     */
    override fun create(repository: ScmProviderRepository, input: PullRequestInput): PullRequest {
        if (StringUtils.isEmpty(input.title)) {
            throw IllegalArgumentException("title cannot be empty")
        }
        if (StringUtils.isEmpty(input.sourceBranch)) {
            throw IllegalArgumentException("source branch cannot be empty")
        }
        if (StringUtils.isEmpty(input.targetBranch)) {
            throw IllegalArgumentException("target branch cannot be empty")
        }

        return TGitApiTemplate.execute(repository, apiFactory) { repo, tGitApi ->
            val builder = TGitMergeRequestParams.builder()
                    .title(input.title)
                    .description(input.body)
                    .sourceBranch(input.sourceBranch)
                    .targetBranch(input.targetBranch)

            if (input.targetRepo is Long) {
                builder.targetProjectId(input.targetRepo as Long)
            }

            val mergeRequest = tGitApi.mergeRequestApi.createMergeRequest(repo.projectIdOrPath, builder.build())
            val targetProject = tGitApi.projectApi.getProject(mergeRequest.targetProjectId)
            val sourceProject = if (mergeRequest.targetProjectId == mergeRequest.sourceProjectId) {
                targetProject
            } else {
                tGitApi.projectApi.getProject(mergeRequest.sourceProjectId)
            }
            TGitObjectConverter.convertPullRequest(mergeRequest, sourceProject, targetProject)
        }
    }

    /**
     * 获取合并请求变更列表
     * @param repository 代码仓库信息
     * @param number 合并请求编号
     * @param opts 列表查询选项
     * @return 变更列表
     */
    override fun listChanges(repository: ScmProviderRepository, number: Int, opts: ListOptions): List<Change> {
        return TGitApiTemplate.execute(repository, apiFactory) { repo, tGitApi ->
            val mergeRequest = getMergeRequestByIid(tGitApi, repo.projectIdOrPath, number)
            val mergeRequestChanges = tGitApi.mergeRequestApi
                    .getMergeRequestChanges(repo.projectIdOrPath, mergeRequest.id)
            mergeRequestChanges.files.map { TGitObjectConverter.convertChange(it) }
        }
    }

    /**
     * 获取合并请求提交列表
     * @param repository 代码仓库信息
     * @param number 合并请求编号
     * @param opts 列表查询选项
     * @return 提交列表
     */
    override fun listCommits(repository: ScmProviderRepository, number: Int, opts: ListOptions): List<Commit> {
        return TGitApiTemplate.execute(repository, apiFactory) { repo, tGitApi ->
            val mergeRequest = getMergeRequestByIid(tGitApi, repo.projectIdOrPath, number)
            tGitApi.mergeRequestApi.getMergeRequestCommits(
                repo.projectIdOrPath,
                mergeRequest.id,
                opts.page,
                opts.pageSize
            ).map { TGitObjectConverter.convertCommit(it) }
        }
    }

    /**
     * 合并指定合并请求
     * @param repository 代码仓库信息
     * @param number 合并请求编号
     */
    override fun merge(repository: ScmProviderRepository, number: Int) {
        TGitApiTemplate.execute(repository, apiFactory) { repo, tGitApi ->
            val mergeRequest = getMergeRequestByIid(tGitApi, repo.projectIdOrPath, number)
            tGitApi.mergeRequestApi.mergeMergeRequest(repo.projectIdOrPath, mergeRequest.id)
        }
    }

    /**
     * 关闭指定合并请求
     * @param repository 代码仓库信息
     * @param number 合并请求编号
     */
    override fun close(repository: ScmProviderRepository, number: Int) {
        TGitApiTemplate.execute(repository, apiFactory) { repo, tGitApi ->
            val mergeRequest = getMergeRequestByIid(tGitApi, repo.projectIdOrPath, number)
            tGitApi.mergeRequestApi.closeMergeRequest(repo.projectIdOrPath, mergeRequest.id)
        }
    }

    /**
     * 获取合并请求评论
     * @param repository 代码仓库信息
     * @param number 合并请求编号
     * @param commentId 评论ID
     * @return 评论详情
     */
    override fun findComment(repository: ScmProviderRepository, number: Int, commentId: Long): Comment {
        return TGitApiTemplate.execute(repository, apiFactory) { repo, tGitApi ->
            val mergeRequest = getMergeRequestByIid(tGitApi, repo.projectIdOrPath, number)
            val note = tGitApi.notesApi.getMergeRequestNote(repo.projectIdOrPath, mergeRequest.id, commentId)
            TGitObjectConverter.convertComment(note)
        }
    }

    /**
     * 获取合并请求评论列表
     * @param repository 代码仓库信息
     * @param number 合并请求编号
     * @param opts 列表查询选项
     * @return 评论列表
     */
    override fun listComments(repository: ScmProviderRepository, number: Int, opts: ListOptions): List<Comment> {
        return TGitApiTemplate.execute(repository, apiFactory) { repo, tGitApi ->
            val mergeRequest = getMergeRequestByIid(tGitApi, repo.projectIdOrPath, number)
            tGitApi.notesApi.getMergeRequestNotes(
                repo.projectIdOrPath,
                mergeRequest.id,
                opts.page,
                opts.pageSize
            ).map { TGitObjectConverter.convertComment(it) }
        }
    }

    /**
     * 创建合并请求评论
     * @param repository 代码仓库信息
     * @param number 合并请求编号
     * @param input 评论输入参数
     * @return 创建的评论详情
     * @throws IllegalArgumentException 当评论内容为空时抛出
     */
    override fun createComment(repository: ScmProviderRepository, number: Int, input: CommentInput): Comment {
        if (StringUtils.isBlank(input.body)) {
            throw IllegalArgumentException("body cannot be empty")
        }
        return TGitApiTemplate.execute(repository, apiFactory) { repo, tGitApi ->
            val mergeRequest = getMergeRequestByIid(tGitApi, repo.projectIdOrPath, number)
            val note = tGitApi.notesApi.createMergeRequestNote(repo.projectIdOrPath, mergeRequest.id, input.body)
            TGitObjectConverter.convertComment(note)
        }
    }

    /**
     * 删除合并请求评论
     * @param repository 代码仓库信息
     * @param number 合并请求编号
     * @param commentId 评论ID
     */
    override fun deleteComment(repository: ScmProviderRepository, number: Int, commentId: Long) {
        // 待实现
    }

    /**
     * 转换合并请求状态枚举
     * @param from 源状态枚举
     * @return 目标状态枚举
     */
    private fun convertFromState(from: PullRequestState?): TGitMergeRequestState? {
        return when (from) {
            PullRequestState.OPENED -> TGitMergeRequestState.OPENED
            PullRequestState.REOPENED -> TGitMergeRequestState.REOPENED
            PullRequestState.MERGED -> TGitMergeRequestState.MERGED
            PullRequestState.CLOSED -> TGitMergeRequestState.CLOSED
            else -> null
        }
    }

    /**
     * 根据编号获取合并请求
     * @param tGitApi TGit API实例
     * @param projectIdOrPath 项目ID或路径
     * @param number 合并请求编号
     * @return 合并请求详情
     * @throws NotFoundScmApiException 当合并请求不存在时抛出
     */
    private fun getMergeRequestByIid(tGitApi: TGitApi, projectIdOrPath: Any, number: Int): TGitMergeRequest {
        return tGitApi.mergeRequestApi.getMergeRequestByIid(projectIdOrPath, number)
            ?: throw NotFoundScmApiException("merge request $number not found")
    }
}
